### PR TITLE
Iot tenant cleaner

### PR DESCRIPTION
**Pull request by ctron**
### Type of change

- Bugfix
- Enhancement / new feature

### Description

Implement the IoT tenant cleanup process for Infinispan. This adds a new finalizer, which will
fork off a new Kubernetes job, that deletes all tenant related information from the Infinispan caches. Once the job is completed successfully, the job gets deleted and the finalizer removed.

### Checklist

- [ ] Update/write design documentation in `./documentation/design`
- [X] Write tests and make sure they pass
- [ ] Update documentation
- [X] Check RBAC rights for Kubernetes / OpenShift roles
- [X] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

### CODE REVIEW

**Comment by ctron** [comment]
This PR also includes PRs: #3426 and #3425. So this PR must be rebased after the other two PRs have been resolved.

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot

File: iot/iot-tenant-cleaner/src/main/java/io/enmasse/iot/tools/cleanup/InfinispanTenantCleaner.java [hunk 1]
```diff
@@ -96,59 +131,61 @@ public void run() throws Exception {
...
+                devicesCache.remove(regKey);
 
             });
 
-            log.info("Deleted {} entries", matches.size());
+            count += matches.size();
+
+            log.info("Deleted {} entries in this iteration", matches.size());
 
         } while (matches.size() > 0);
```
Comment (jbtrystram) [inline iot/iot-tenant-cleaner/src/main/java/io/enmasse/iot/tools/cleanup/InfinispanTenantCleaner.java]: This result in an extra hotrod request + server query which may be costly. Testing `matches.size()=DELETION_CHUNK_SIZE` avoid that last call

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot

File: pkg/controller/iotproject/tenant_cleanup.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	iotv1alpha1 "github.com/enmasseproject/enmasse/pkg/apis/iot/v1alpha1"
+)
+
+var (
+	tenantCleanupNamespace uuid.UUID = uuid.MustParse("20c0db1c-ffc8-11e9-800c-c85b762e5a2c")
```
[the remark this answers is not in the document]
Comment (ctron) [reply pkg/controller/iotproject/tenant_cleanup.go]: This should ensure the uniqueness for a specific use case when combining this with a name based UUID. I think using a random UUID as the namespace is valid approach.

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.shared.http.HttpAdapterTest

File: iot/iot-tenant-cleaner/src/main/resources/logback.xml [hunk 3]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<!DOCTYPE xml>
+
+<configuration>
+
+  <appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">
+    <encoder>
+      <pattern>%d{HH:mm:ss.SSS} [%thread] %-5level %logger{36} - %msg%n</pattern>
```
Comment (lulf) [inline iot/iot-tenant-cleaner/src/main/resources/logback.xml]: We should standardize on the log format for all components. This just caught my attention now, so maybe its a separate PR to fix in other components as well. I believe what we have standardized on (for both java and node at least), is:

```
<pattern>%d{yyyy-MM-dd'T'HH:mm:ss.SSS'Z',GMT} %-5p [%c{0}] %m%n</pattern>
```

File: templates/iot/operator/020-Role-iot-operator.yaml [hunk 4]
```diff
@@ -0,0 +1,13 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
```
Comment (lulf) [inline templates/iot/operator/020-Role-iot-operator.yaml]: This role should be added to the CSV in olm-manifest/src/main/resources/latest/... as well

File: iot/iot-tenant-cleaner/src/main/resources/logback.xml (hunk 3 above)
Comment (ctron) [reply iot/iot-tenant-cleaner/src/main/resources/logback.xml]: Yes, I only have that in there for debugging. Not having the thread name can be a problem. I will remove that.

File: templates/iot/operator/020-Role-iot-operator.yaml (hunk 4 above)
Comment (ctron) [reply templates/iot/operator/020-Role-iot-operator.yaml]: :+1: Done.

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.shared.http.HttpAdapterTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.shared.http.HttpAdapterTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.shared.http.HttpAdapterTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.shared.http.HttpAdapterTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.shared.http.HttpAdapterTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.shared.http.HttpAdapterTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.shared.http.HttpAdapterTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.shared.http.HttpAdapterTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.shared.http.HttpAdapterTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.shared.http.HttpAdapterTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.shared.http.HttpAdapterTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.shared.http.HttpAdapterTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.shared.http.HttpAdapterTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.shared.http.HttpAdapterTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.shared.http.HttpAdapterTest

**Comment by ctron** [comment]
This is now part of #3556

**Pull request closed without merging**